File: platforms/android/sdk/src/test/kotlin/dev/bota/sdk/ClientPresenceTest.kt
package dev.bota.sdk

import java.io.File
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class ClientPresenceTest {
    @Test
    fun canonicalLifecycle() {
        val root = generateSequence(File(checkNotNull(System.getProperty("user.dir")))) { it.parentFile }
            .first { File(it, "protocol/client-presence/v1.json").isFile }
        val fixture = JSONObject(File(root, "protocol/client-presence/v1.json").readText())
        val sessions = fixture.getJSONArray("sessions")
        val presence = ConnectionClientPresence()
        val tokens = mutableMapOf<String, String>()
        var nextSession = 0
        val steps = fixture.getJSONArray("steps")
        for (index in 0 until steps.length()) {
            val step = steps.getJSONObject(index)
            when (step.getString("action")) {
                "verified_connect" -> {
                    val id = sessions.getString(nextSession.coerceAtMost(sessions.length() - 1))
                    tokens[step.getString("connection")] = id
                    presence.connected(step.getString("device_id"), id, id)
                    nextSession++
                }
                "disconnect" -> presence.disconnected(tokens.getValue(step.getString("connection")))
                "destroy" -> presence.destroy()
                "report" -> {
                    val report = presence.nextReport(step.getString("device_id"))
                    if (step.isNull("expected")) assertNull(report) else {
                        val expected = step.getJSONObject("expected")
                        assertEquals(sessions.getString(expected.getInt("session")), report?.sessionId)
                        assertEquals(expected.getLong("sequence"), report?.sequence)
                        assertEquals(SdkIdentity.PACKAGE, report?.sdkPackage)
                        assertEquals(SdkIdentity.VERSION, report?.sdkVersion)
                    }
                }
                else -> error("unknown fixture action")
            }
        }
    }
}

File: frameworks/web/src/__tests__/clientPresence.test.ts
import assert from 'node:assert/strict'
import { readFile } from 'node:fs/promises'
import test from 'node:test'
import { ConnectionClientPresence } from '../clientPresence.ts'

test('client presence follows the shared passive lifecycle fixtures without HTTP', async t => {
  const fixture = JSON.parse(await readFile(new URL('../../../../protocol/client-presence/v1.json', import.meta.url), 'utf8'))
  const pkg = JSON.parse(await readFile(new URL('../../package.json', import.meta.url), 'utf8'))
  const fetch = t.mock.method(globalThis, 'fetch', async () => { throw new Error('metadata must not fetch') })
  let next = 0
  const owner = new ConnectionClientPresence(() => fixture.sessions[next++])
  const disconnects = new Map<string, () => void>()
  for (const step of fixture.steps) {
    if (step.action === 'verified_connect') disconnects.set(step.connection, owner.connected(step.device_id))
    else if (step.action === 'disconnect') disconnects.get(step.connection)!()
    else if (step.action === 'destroy') owner.destroy()
    else {
      const report = await owner.nextReport(step.device_id)
      if (step.expected === null) assert.equal(report, null)
      else {
        assert.deepEqual(report, {
          schema_version: 1, platform: 'web', sdk_package: '@bota.dev/web-app-sdk',
          sdk_version: pkg.version, session_id: fixture.sessions[step.expected.session],
          sequence: step.expected.sequence,
        })
      }
    }
  }
  assert.equal(fetch.mock.callCount(), 0)
})
